File: packages/core/src/styles/ghosts.ts
import { fnv1a, seeded } from '../hash';
import { extractInitials } from '../initials';
import { GHOST_BGS, GHOST_CHEEKS, GHOST_COLORS } from '../palettes';
import type { GhostsSpec } from '../types';

export function ghostsStyle(name: string): GhostsSpec {
	const initials = extractInitials(name);
	const h = fnv1a(name.trim().toLowerCase());

	const bg = GHOST_BGS[h % GHOST_BGS.length];
	const color = GHOST_COLORS[seeded(h, 1) % GHOST_COLORS.length];
	const cheek = GHOST_CHEEKS[seeded(h, 1) % GHOST_CHEEKS.length];

	const bodyW = 28 + (seeded(h, 2) % 8);
	const bodyTop = 22 + (seeded(h, 3) % 6);
	const bodyBottom = 72 + (seeded(h, 4) % 6);
	const numWaves = 3 + (seeded(h, 5) % 3);
	const waveAmp = 4 + (seeded(h, 6) % 4);

	const eyeY = bodyTop + bodyW * 0.6 + 2;
	const eyeSpacing = 7 + (seeded(h, 8) % 4);
	const mouthY = eyeY + 8;

	const numSparkles = 2 + (seeded(h, 13) % 3);
	const sparkles: Array<{ cx: number; cy: number; r: number }> = [];
	for (let i = 0; i < numSparkles; i++) {
		sparkles.push({
			cx: 15 + (seeded(h, i + 30) % 70),
			cy: 15 + (seeded(h, i + 40) % 70),
			r: 1 + (seeded(h, i + 50) % 2),
		});
	}

	return {
		style: 'ghosts',
		initials,
		palette: { bg, color, cheek },
		bodyW,
		bodyTop,
		bodyBottom,
		numWaves,
		waveAmp,
		eyeType: seeded(h, 7) % 5,
		eyeY,
		eyeSpacing,
		mouthType: seeded(h, 9) % 4,
		mouthY,
		hasBlush: seeded(h, 10) % 2 === 0,
		floatY: -3 + (seeded(h, 11) % 7),
		tilt: -3 + (seeded(h, 12) % 7),
		sparkles,
	};
}

File: packages/dart/lib/src/styles/ghosts.dart
import '../hash.dart';
import '../initials.dart';
import '../palettes.dart';
import '../types.dart';

GhostsSpec ghostsStyle(String name) {
  final initials = extractInitials(name);
  final h = fnv1a(name.trim().toLowerCase());

  final bg = ghostBgs[h % ghostBgs.length];
  final color = ghostColors[seeded(h, 1) % ghostColors.length];
  final cheek = ghostCheeks[seeded(h, 1) % ghostCheeks.length];

  final bodyW = 28 + (seeded(h, 2) % 8);
  final bodyTop = 22 + (seeded(h, 3) % 6);
  final bodyBottom = 72 + (seeded(h, 4) % 6);
  final numWaves = 3 + (seeded(h, 5) % 3);
  final waveAmp = 4 + (seeded(h, 6) % 4);

  final eyeY = bodyTop + bodyW * 0.6 + 2;
  final eyeSpacing = 7 + (seeded(h, 8) % 4);
  final mouthY = eyeY + 8;

  final numSparkles = 2 + (seeded(h, 13) % 3);
  final sparkles = <Sparkle>[];
  for (int i = 0; i < numSparkles; i++) {
    sparkles.add(Sparkle(
      cx: 15 + (seeded(h, i + 30) % 70),
      cy: 15 + (seeded(h, i + 40) % 70),
      r: 1 + (seeded(h, i + 50) % 2),
    ));
  }

  return GhostsSpec(
    initials: initials,
    palette: GhostsPalette(bg: bg, color: color, cheek: cheek),
    bodyW: bodyW,
    bodyTop: bodyTop,
    bodyBottom: bodyBottom,
    numWaves: numWaves,
    waveAmp: waveAmp,
    eyeType: seeded(h, 7) % 5,
    eyeY: eyeY,
    eyeSpacing: eyeSpacing,
    mouthType: seeded(h, 9) % 4,
    mouthY: mouthY,
    hasBlush: seeded(h, 10) % 2 == 0,
    floatY: -3 + (seeded(h, 11) % 7),
    tilt: -3 + (seeded(h, 12) % 7),
    sparkles: sparkles,
  );
}
